File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_count = 0
        n = len(s)
        i, j = 0, 0
        
        visited = set()

        while j < n:
            if s[j] not in visited:
                visited.add(s[j])
                max_count = max(max_count, j-i+1)
            else:
                max_count = max(max_count, j-i)
                
                while i < j:
                    if s[i] == s[j]:
                        i += 1
                        break
                        
                    visited.remove(s[i])
                    i += 1
                
            j += 1
                        
        return max_count

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int longest = 0;
        int i = 0, j = 0;
        
        HashMap<Character, Integer> map = new HashMap();
        
        while (j < s.length()) {
            
            if (map.containsKey(s.charAt(j))) {
                i = Math.max(map.get(s.charAt(j)) + 1, i) ;
            } 
            
            map.put(s.charAt(j), j);            
            longest = Math.max(longest, j - i + 1);
            j += 1;
        }
        
        return longest;
    }
}